test/framework/universal_app.go

```diff
@@ -263,27 +265,44 @@ func NewUniversalApp(t testing.TestingT, clusterName, dpName, mesh string, mode
 	return app, nil
 }
 
-func (s *UniversalApp) allocatePublicPortsFor(ports ...string) {
-	for _, port := range ports {
-		pubPortUInt32, err := util_net.PickTCPPort("", s.lastUsedPort+1, 11204)
+func (s *UniversalApp) updatePublishedPorts() error {
+	var ports []uint32
+	for portStr := range s.ports {
+		port, err := strconv.Atoi(portStr)
 		if err != nil {
-			panic(err)
+			return err
 		}
-		s.ports[port] = strconv.Itoa(int(pubPortUInt32))
-		s.lastUsedPort = pubPortUInt32
+		ports = append(ports, uint32(port))
+	}
+
+	publishedPorts, err := GetPublishedDockerPorts(s.t, s.container, ports)
+	if err != nil {
+		return err
+	}
+
+	for _, port := range ports {
+		publishedPort := publishedPorts[port]
+		s.ports[strconv.Itoa(int(port))] = strconv.Itoa(int(publishedPort))
+	}
+	return nil
+}
+
+func (s *UniversalApp) allocatePublicPortsFor(ports ...string) {
+	for _, port := range ports {
+		s.ports[port] = "0"
 	}
 }
 
 func (s *UniversalApp) publishPortsForDocker(isipv6 bool) (args []string) {
 	// If we aren't using IPv6 in the container then we only want to listen on
 	// IPv4 interfaces to prevent resolving 'localhost' to the IPv6 address of
 	// the container and having the container not respond.
-	ip := "0.0.0.0:"
+	ip := "0.0.0.0::"
```

Comment:
The double "::" is expected?

Reply:
yes, it's a weird format of "give me random port but bind only to 0.0.0.0"